File: GradleMC/Forge/Minecraft 1.20.1/src/main/java/com/soumyajit/gradlemc/client/overlay/OverlayLineComposer.java
package com.soumyajit.gradlemc.client.overlay;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/** Pure title/FPS composition so presentation choices are testable without a Minecraft client. */
public final class OverlayLineComposer {
    private OverlayLineComposer() {
    }

    public static List<String> compose(boolean compact, boolean showTitle, boolean showFps, boolean showAverageFps,
                                       FpsRollingStatsCalculator.Snapshot fps) {
        List<String> lines = new ArrayList<>();
        if (compact) {
            List<String> parts = new ArrayList<>();
            if (showTitle) {
                parts.add("GradleMC");
            }
            if (showFps) {
                parts.add(whole(fps.currentFps()) + " FPS");
            }
            if (showAverageFps) {
                parts.add("avg " + whole(fps.averageFps()));
            }
            if (!parts.isEmpty()) {
                lines.add(String.join(" | ", parts));
            }
            return List.copyOf(lines);
        }
        if (showTitle) {
            lines.add("GradleMC");
        }
        if (showFps) {
            lines.add("FPS: " + whole(fps.currentFps()));
        }
        if (showAverageFps) {
            lines.add("Average FPS: " + whole(fps.averageFps()));
        }
        return List.copyOf(lines);
    }

    private static String whole(double value) {
        return String.format(Locale.ROOT, "%.0f", Double.isFinite(value) ? Math.max(0.0D, value) : 0.0D);
    }
}

File: GradleMC/Forge/Minecraft 1.20.1/src/main/kotlin/com/soumyajit/gradlemc/client/overlay/OverlayLineComposer.kt
package com.soumyajit.gradlemc.client.overlay

import java.util.Locale

/** Pure overlay text composition; each visible item is controlled independently. */
internal object OverlayLineComposer {
    fun compose(
        showTitle: Boolean,
        showCurrentFps: Boolean,
        showAverageFps: Boolean,
        currentFps: Double?,
        averageFps: Double?,
    ): List<String> = buildList(3) {
        if (showTitle) add("GradleMC")
        if (showCurrentFps) add("FPS: ${format(currentFps)}")
        if (showAverageFps) add("Average FPS: ${format(averageFps)}")
    }

    private fun format(value: Double?): String =
        if (value == null || !value.isFinite()) {
            "warming up"
        } else {
            String.format(Locale.ROOT, "%.0f", value.coerceAtLeast(0.0))
        }
}
